### PR TITLE
feat(graphql): Additional trigger filtering

### PR DESCRIPTION
**Pull request by robzienert**
Here's some more stuffs for our POC. Starting to find my feet from a server perspective; the client side of things is still quite dodgy.

```graphql
query {
  pipelines(havingTriggerType: "cron", disabled: false) {
    id
    name
    application
    triggers(types: ["cron"]) {
      ... on CronTrigger {
        id
        enabled
        cronExpression
      }
    }
  }
}
```

### CODE REVIEW

File: front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/Trigger.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.google.common.collect.ForwardingMap;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+public class Trigger extends ForwardingMap<String, Object> {
+
+  public final static TypeReference<Trigger> TYPE = new TypeReference<Trigger>() {};
```
Comment (dreynaud) [inline front50-core/src/main/groovy/com/netflix/spinnaker/front50/model/pipeline/Trigger.java]: not used?

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java [hunk 2]
```diff
@@ -42,18 +39,44 @@ public PipelinesDataFetcher(PipelineDAO pipelineDAO) {
...
+      super(environment);
+    }
+
+    @Override
+    public boolean test(Pipeline pipeline) {
+      if (!environment.containsArgument("disabled")) {
+        return true;
+      }
+      boolean disabled = environment.getArgument("disabled");
+      return (Boolean) pipeline.getOrDefault("disabled", false) == disabled;
```
Comment (dreynaud) [inline front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java]: I don't think the cast to `(Boolean)` is needed

Comment (robzienert) [reply front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java]: It is needed. `Pipeline` is a `HashMap<String, Object>`, so even if the default value is a `boolean`, the returned type of `getOrDefault` will still be `Object`.

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java [hunk 3]
```diff
@@ -42,18 +39,44 @@ public PipelinesDataFetcher(PipelineDAO pipelineDAO) {
...
-    return pipelines;
+  static class DisabledPredicate extends DataFetchingEnvironmentPredicate<Pipeline> {
+    DisabledPredicate(DataFetchingEnvironment environment) {
+      super(environment);
+    }
+
+    @Override
+    public boolean test(Pipeline pipeline) {
+      if (!environment.containsArgument("disabled")) {
+        return true;
```
[the remark this answers is not in the document]
Comment (robzienert) [reply front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java]: I will change the logic.

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java (hunk 2 above)
Comment (dreynaud) [reply front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java]: oh boy, I misinterpreted the operator precedence here. I thought the case applied to the whole `==` expression, not just the left hand side.

File: front50-web/src/test/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/TriggersDataFetcherSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    when:
+    def result = subject.get(environment)
+
+    then:
+    result == triggers
+
+    where:
+    triggers = [
+      new Trigger([type: "cron"])
```
[the remark this answers is not in the document]
Comment (robzienert) [reply front50-web/src/test/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/TriggersDataFetcherSpec.groovy]: Derp.

File: front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (robzienert) [reply front50-web/src/main/groovy/com/netflix/spinnaker/front50/graphql/datafetcher/PipelinesDataFetcher.java]: Added a new test for disabled filtering specifically. Behaves more reasonably now.

**Pull request merged**